File: problems/501-find-mode-in-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        count = {}
        max_count = 0
        
        def inorder(node):
            nonlocal max_count
            if node:
                inorder(node.left)
                if node.val in count:
                    count[node.val] += 1
                else:
                    count[node.val] = 1
                max_count = max(max_count, count[node.val])
                inorder(node.right)
        
        inorder(root)
        
        result = []
        for val, c in count.items():
            if c == max_count:
                result.append(val)
        return result

File: problems/501-find-mode-in-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxCount = 0;
    int currCount = 0;
    int prevVal = Integer.MIN_VALUE;
    List<Integer> modes = new ArrayList<>();

    public int[] findMode(TreeNode root) {
        inorder(root);
        int[] result = new int[modes.size()];
        for (int i = 0; i < modes.size(); i++) {
            result[i] = modes.get(i);
        }
        return result;
    }

    private void inorder(TreeNode node) {
        if (node == null) return;
        inorder(node.left);
        if (node.val == prevVal) {
            currCount++;
        } else {
            currCount = 1;
            prevVal = node.val;
        }
        if (currCount > maxCount) {
            maxCount = currCount;
            modes.clear();
            modes.add(node.val);
        } else if (currCount == maxCount) {
            modes.add(node.val);
        }
        inorder(node.right);
    }
}